interpreter/script/run.ml

```diff
@@ -248,18 +248,22 @@ let string_of_nan = function
   | CanonicalNan -> "nan:canonical"
   | ArithmeticNan -> "nan:arithmetic"
 
-let type_of_result r =
-  match r with
+let rec type_of_result r =
+  match r.it with
   | LitResult v -> Values.type_of v.it
   | NanResult n -> Values.type_of n.it
+  | EitherResult rs -> type_of_result (List.hd rs)
```

Comment:
There appears to be some jankiness here - for example, if I change the `either` line in `thread.wast` to
```
  (assert_return (invoke "run") (either (i32.const 0) (i64.const 42)))
```

The test will randomly either pass, or fail with the message
```
> Result: 42 : i32
> Expect: (0 | 42) : i32
> -- Error: 
> ../test/core/thread.wast:24.3-24.71: assertion failure: wrong return values
```

Not sure what the right approach to polish this is... force all elements of `either` to have the same type?

Reply:
You are picky. :) You can get the same error without using `either`. The script level is dumb and does no type checking (which actually is a feature, since the interpreter can intentionally run unvalidated code with -u). What's a bit misleading here, though, is the random type annotation that gets printed. I extended that to check if the either types are consistent and print "?" otherwise. Though once we merge upstream, this will require computing a lub to be accurate, so not sure we'll really want to keep bothering...

Reply:
I think I'm comfortable being imprecise with `?` (even when there's a LUB) rather than printing the wrong type - change makes sense!